src/proposer.ts

```diff
@@ -121,6 +130,246 @@ let s: ReturnType<typeof strings>
+ * - For each index i: asset/amount/chain_id must match between input and output
+ * - outputs[i].to must be provided (no to_external) and must be a valid on-chain vault ID
+ * - Fees must be zero:
+ *   - totalFee amounts must all be "0"
+ *   - proposerTip must be empty
+ *   - protocolFee must be empty
+ *   - agentTip must be undefined or empty
+ */
+
+export async function validateAssignDepositStructure(
+	intention: Intention
+): Promise<void> {
+	if (!Array.isArray(intention.inputs) || !Array.isArray(intention.outputs)) {
+		throw new Error('AssignDeposit requires inputs and outputs arrays')
+	}
+	if (intention.inputs.length !== intention.outputs.length) {
+		throw new Error(
+			'AssignDeposit requires 1:1 mapping between inputs and outputs'
+		)
+	}
+
+	// Zero-fee enforcement
+	if (!Array.isArray(intention.totalFee) || intention.totalFee.length === 0) {
+		throw new Error('AssignDeposit requires totalFee with zero amount')
+	}
+	const allTotalZero = intention.totalFee.every((f) => f.amount === '0')
+	if (!allTotalZero) {
+		throw new Error('AssignDeposit totalFee must be zero')
+	}
+	if (
+		Array.isArray(intention.proposerTip) &&
+		intention.proposerTip.length > 0
+	) {
+		throw new Error('AssignDeposit proposerTip must be empty')
+	}
+	if (
+		Array.isArray(intention.protocolFee) &&
+		intention.protocolFee.length > 0
+	) {
+		throw new Error('AssignDeposit protocolFee must be empty')
+	}
+	if (Array.isArray(intention.agentTip) && intention.agentTip.length > 0) {
+		throw new Error('AssignDeposit agentTip must be empty if provided')
+	}
+
+	for (let i = 0; i < intention.inputs.length; i++) {
+		const input: IntentionInput = intention.inputs[i]
+		const output: IntentionOutput = intention.outputs[i]
+
+		if (!output || (output.to === undefined && !output.to_external)) {
+			throw new Error('AssignDeposit requires outputs[].to (vault ID)')
+		}
+		if (output.to_external !== undefined) {
+			throw new Error('AssignDeposit does not support to_external')
+		}
+
+		if (input.asset.toLowerCase() !== output.asset.toLowerCase()) {
+			throw new Error('AssignDeposit input/output asset mismatch at index ' + i)
+		}
+		if (input.amount !== output.amount) {
+			throw new Error(
+				'AssignDeposit input/output amount mismatch at index ' + i
+			)
+		}
+		if (input.chain_id !== output.chain_id) {
+			throw new Error(
+				'AssignDeposit input/output chain_id mismatch at index ' + i
+			)
+		}
+
+		// Validate on-chain vault existence
+		await validateVaultIdOnChain(Number(output.to))
+	}
+}
+
+/**
+ * Discovers ERC-20 deposits made by `controller` into the VaultTracker and
+ * ingests them into the local `deposits` table. Uses Alchemy's decoded
+ * getAssetTransfers API for reliability and simplicity.
+ *
+ * If fromBlock/toBlock are not provided, computes a default lookback window
+ * of ~7 days by subtracting ~50,400 blocks from the latest block.
+ */
+async function discoverAndIngestErc20Deposits(params: {
+	controller: string
```

Comment:
See earlier note that I think we can leave `controller` unspecified and get all deposits in the time range for a given `token`.